File: app/src/main/java/com/giaquino/android/cropped/data/repository/TopicRepositoryImpl.java
package com.giaquino.android.cropped.data.repository;

import com.giaquino.android.cropped.data.api.UnsplashApi;
import com.giaquino.android.cropped.data.common.Resource;
import com.giaquino.android.cropped.data.model.Topic;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Single;

public class TopicRepositoryImpl implements TopicRepository {

    private UnsplashApi api;

    private List<Topic> topics = new ArrayList<>();

    @Inject
    public TopicRepositoryImpl(UnsplashApi api) {
        this.api = api;
    }

    @Override
    public Single<Resource<List<Topic>>> getTopics(String orderBy, int page, int perPage) {
        if (!topics.isEmpty()) {
            return Single.just(Resource.success(topics));
        }
        return api.getTopics(orderBy, page, perPage)
                .map(Resource::success)
                .onErrorReturn(Resource::failure)
                .doOnSuccess(resource -> resource.isSuccess(data -> {
                    topics.clear();
                    topics.addAll(data);
                }));
    }
}

File: app/src/main/java/com/giaquino/android/cropped/data/repository/TopicRepositoryImpl.kt
package com.giaquino.android.cropped.data.repository

import com.giaquino.android.cropped.data.api.UnsplashApi
import com.giaquino.android.cropped.data.common.Resource
import com.giaquino.android.cropped.data.model.Topic
import io.reactivex.rxjava3.core.Single
import java.util.*
import java.util.function.Consumer
import javax.inject.Inject

class TopicRepositoryImpl @Inject constructor(private val api: UnsplashApi) : TopicRepository {

    private val topics: MutableList<Topic> = ArrayList()

    override fun getTopics(orderBy: String, page: Int, perPage: Int): Single<Resource<List<Topic>>> {
        return when {
            topics.isNotEmpty() -> {
                Single.just(Resource.success<List<Topic>>(topics))
            }
            else -> {
                api.getTopics(orderBy, page, perPage)
                        .map<Resource<List<Topic>>> { data: List<Topic> -> Resource.success(data) }
                        .onErrorReturn { error: Throwable -> Resource.failure(error) }
                        .doOnSuccess { resource: Resource<List<Topic>> ->

                            resource.isSuccess { data ->
                                topics.clear()
                                topics.addAll(data)
                            }
                        }
            }
        }
    }
}